dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,567 @@
+
+    def get_metadata_by_id(self, file_id, **kwargs):
+        return self.request(
+            "GET", "drive/v3/files/" + file_id, **kwargs
+        ).json()
+
+    def search(self, parent=None, name=None, add_params={}):
+        query = []
+        if parent is not None:
+            query.append("'{}' in parents".format(parent))
+        if name is not None:
+            query.append("name = '{}'".format(name))
+        params = {"q": " and ".join(query), "spaces": self.space}
+        params.update(add_params)
+        while True:
+            data = self.request("GET", "drive/v3/files", params=params).json()
+            for i in data["files"]:
+                yield i
+            if not data.get("nextPageToken"):
+                break
+            params["pageToken"] = data["nextPageToken"]
+
+    def get_metadata_by_path(self, path, root=None, fields=[]):
+        if root is None:
+            root = self.root
+        parent = self.get_metadata_by_id(root)
+        current_path = [parent["name"]]
+        parts = path.split("/")
+        # only specify fields for the last search query
+        kwargs = [{}] * (len(parts) - 1) + [
+            {"add_params": {"fields": "files({})".format(",".join(fields))}}
+            if fields
+            else {}
+        ]
+        for part, kwargs in zip(parts, kwargs):
+            if not self.metadata_isdir(parent):
+                raise GDriveError(
+                    "{} is not a folder".format("/".join(current_path))
+                )
+            current_path.append(part)
+            files = list(self.search(parent["id"], part, **kwargs))
+            if len(files) > 1:
+                raise GDriveError(
+                    "path {} is duplicated".format("/".join(current_path))
+                )
+            elif len(files) == 0:
+                raise GDriveResourceNotFound("/".join(current_path))
+            parent = files[0]
+        return parent
+
+    def metadata_isdir(self, metadata):
+        return metadata["mimeType"] == self.MIME_GOOGLE_APPS_FOLDER
+
+    def get_file_checksum(self, path_info):
+        metadata = self.get_metadata_by_path(
+            path_info.path,
+            root=path_info.root,
+            params={"fields": "md5Checksum"},
+        )
+        return metadata["md5Checksum"]
+
+    def _list_files(self, folder_id):
+        for i in self.search(parent=folder_id):
+            if self.metadata_isdir(i):
+                for j in self._list_files(i["id"]):
+                    yield i["name"] + "/" + j
+            else:
+                yield i["name"]
+
+    def list_cache_paths(self):
+        return self._list_files(self.root)
+
+    def mkdir(self, parent, name):
+        data = {
+            "name": name,
+            "mimeType": self.MIME_GOOGLE_APPS_FOLDER,
+            "parents": [parent],
+            "spaces": self.space,
+        }
+        return self.request("POST", "drive/v3/files", json=data).json()
+
+    @only_once
```

Comment:
Could you please explain why this is needed?

Reply:
It is needed to fix duplicated directories creation. DVC runs several threads for push process, and they could try to create this directories simultaneously.

Reply:
But this way it wouldn't be possible to remove the directory and create it again. Hm...

Reply:
But it is not a problem since we don't delete anything as of yet...
@efiop ?

Reply:
Remotes don't delete directories made for cache, indeed. Shouldn't be a problem.